tonconnect-server/TonConnect2.md

```diff
@@ -0,0 +1,533 @@
+
+Access to TON apps is controlled by the wallets: regular applications that keep users’ cryptographic keys on their devices. Wallets enable users to sign into applications and authorize blockchain transactions.
+
+The goal of Ton Connect is to enable smooth and safe interaction between wallets, services and apps.
+
+
+### Apps vs services
+
+In this specification we use the following terminology:
+
+**App** is a software that runs on behalf of the user and does not rely on a dedicated backend server for accessing users’ data.
+
+**Service** is a software that runs on behalf of its operator, stores users’ data and regulates access to the data.
+
+For TON Connect the distinction lies in the fact that service has its own offchain identity and may request [identification of the user](#user-identification). Apps run on behalf of the user, so identifying the user does not make much sense, while app’s own identity is defined by its smart contract address.
+
+
+### Decentralized vs centralized apps
+
+In the above definitions apps are considered decentralized and services considered centralized. 
+In reality, there is a spectrum of trust-minimization and each specific app or service may rely on some centralized infrastructure.
+
+Both apps and services are often non-custodial for the security and legal reasons. Instead of managing cryptographic keys that hold users’ funds they rely on wallet apps to approve and sign transactions on behalf of the users. Wallet apps themselves are often non-custodial too: they store secret keys securely on their users’ devices and never communicate them to other computers on the network.
+
+
+### User identification
+
+To minimize cross-service tracking, TON Connect uses distinct cryptographic identities for each service.
+
+User identity is tied to the [app keypair](#app-keypair).
+
+
+### User authentication
+
+Decentralized apps authenticate users within smart contracts using message sender addresses. 
+The UI (“offchain logic”) runs on behalf of the user, so there is no need for a dedicated “sign in” scheme.
+
+Centralized apps run on the servers and may store user data. They may use “sign in” functionality to authenticate the user, where instead of username/password pair we use cryptographic keypair derived from the wallet secret seed.
+
+
+### App authentication
+
+Apps are authenticated using [TON DNS](https://ton.org/docs/#/web3/dns) protocol.
+
+Decentralized apps may use `sha256("wallet")` key for the smart contract address. 
+When the wallet receives request to sign a message to a smart contract with a given **.ton** name, a valid and up-to-date DNS record would be used to authenticate that contract address with that name.
+
+**TODO:** callback/return URLs should also be specified in the TON DNS to permit their verification for dapps that cannot sign their own requests.
```

Comment:
Another question about (decentralized) apps. What if the app has a collection of smart contracts that should be interacted with. What's the recommended best practice in this case? Add a subdomain for each of them and a sha256("wallet") key under each subdomain? I didn't dig into the DNS standard yet so I'm not sure how easy this is. If this is the case, this should be explained.

Reply:
If you have multiple entry points to your dapp that don't need separate branding, then enumerating addresses would be sufficient, i think. We could think of a chained list of addresses similar to [TIP64](https://github.com/ton-blockchain/TIPs/issues/64). Except that we could fit up to 3 addresses per cell instead of 1 as in TIP64.